File: lib/filter/low_pass_filter.dart
import '../native_cpp.dart';
import 'dart:ffi'; // For FFI

/// T curOutput = alpha_ * curInput + (1.0 - alpha_) * prevOutput_;
final double Function(double curInput) lowPassFilterCalc = cppRoboticsLib
    .lookup<NativeFunction<Double Function(Double)>>('LowPassFilterCalc')
    .asFunction();

/// params[0] = config.cutOffFreqHz;
///
/// params[1] = config.timeConstantSec;
final Pointer<Double> Function() lowPassFilterGetParams = cppRoboticsLib
    .lookup<NativeFunction<Pointer<Double> Function()>>(
        'LowPassFilterGetParams')
    .asFunction();

final void Function(double cutOffFreqHz, double timeConstantSec)
    lowPassFilterSetParams = cppRoboticsLib
        .lookup<NativeFunction<Void Function(Double, Double)>>(
            'LowPassFilterSetParams')
        .asFunction();

final void Function() lowPassFilterResetFilter = cppRoboticsLib
    .lookup<NativeFunction<Void Function()>>('LowPassFilterResetFilter')
    .asFunction();

final void Function() lowPassFilterResetPrevValues = cppRoboticsLib
    .lookup<NativeFunction<Void Function()>>('LowPassFilterResetPrevValues')
    .asFunction();

File: cppRobotics/extern/filter/LowPassFilter.cpp
#include "LowPassFilter.h"

static LowPassFilter<double> lowPassFilter(50);
static double pParams[2];

double LowPassFilterCalc(double curInput) {
  return lowPassFilter.Calc(curInput);
}

double* LowPassFilterGetParams() {
  FirstOrderFilterConfig config = lowPassFilter.GetParams();
  pParams[0] = config.cutOffFreqHz;
  pParams[1] = config.timeConstantSec;
  return pParams;
}

void LowPassFilterSetParams(double cutOffFreqHz, double timeConstantSec) {
  FirstOrderFilterConfig config = {cutOffFreqHz, timeConstantSec};
  lowPassFilter.SetParams(config);
}

void LowPassFilterResetFilter() {
  lowPassFilter.ResetFilter();
}

void LowPassFilterResetPrevValues() {
  lowPassFilter.ResetPrevValues();
}
